counterfactuals/environment.py

```diff
@@ -0,0 +1,133 @@
+import babyai
+import pickle
+from gym_minigrid.minigrid import MiniGridEnv, OBJECT_TO_IDX, COLOR_TO_IDX
+import gym
+import numpy as np
+import random
+from dreamer.envs.wrapper import EnvWrapper
+from dreamer.envs.one_hot import OneHotAction
+from gym.spaces import Box
+
+class SlipperinessWrapper(EnvWrapper):
+    def __init__(self, env, slip_prob):
+        super().__init__(env)
+        self.slip_prob = slip_prob
+
+    def step(self, action):
+        if action == MiniGridEnv.Actions.forward: # Go forward action; no slippage for other actions
+            if np.random.uniform() < self.slip_prob:
+                action = random.choice(["left", "right"])
+                # By default, the agent can only move in a direction if it's facing that way.
+                # We can model slippage by turning a direction, moving that way, then turning back
+                # and only counting it as a single action.
+                self.step_count -= 2
+                if action == "left":
+                    super().step(MiniGridEnv.Actions.left)
+                    super().step(MiniGridEnv.Actions.forward)
+                    obs = super().step(MiniGridEnv.Actions.right)
+                else:
+                    super().step(MiniGridEnv.Actions.right)
+                    super().step(MiniGridEnv.Actions.forward)
+                    obs = super().step(MiniGridEnv.Actions.left)
+                return obs
+        return super().step(action)
+
+
+class OneHotObs(EnvWrapper):
+
+    def __init__(self, env):
+        super().__init__(env)
+        example_obs = self.reset()
+        obs_space = {}
+        for k, v in example_obs.items():
+            v_shape = v * 0
+            obs_space[k] = Box(v_shape, v_shape + 1, dtype=np.int32)
+        self.obs_space = obs_space
+
+    def step(self, action):
+        obs, reward, done, info = self.env.step(action)
+        obs = self.one_hot(obs)
+        return obs, reward, done, info
+
+    def reset(self):
+        obs = self.env.reset()
+        return self.one_hot(obs)
+
+    def one_hot(self, obs):
+
+        one_hot_obs = {}
+
+        # One-hotify direction
+        NUM_DIRECTIONS = 4
+        dir = np.zeros(NUM_DIRECTIONS)
+        dir[obs["direction"]] = 1
+        one_hot_obs["direction"] = dir
+
+        # One-hotify mission
+        # TODO: This is tricky!  Missions can be arbitrary lengths, so we could either
+        #       (a) One-hot encode each token sequentially, return a variable-length array
+        #       (b) Give a fix-sized mission vector where smaller missions are zero-padded
+        #       (c) Assume we'll only use levels with a particular mission structure
+        #       The code currently does (c).  We assume there's at most one object mentioned.
+        #       But at some point we should probably switch to the more generalizable version using the babyai
+        #       repo's preprocessor InstructionsPreprocessor in file utils/format.py
```

Comment:
What do you mean by mission? Shouldn't every step just return 1 set of obs of fixed size, or what am I missing?

Reply:
The observation is a dictionary with multiple keys, including the fixed-size grid andThe mission is something like "go to the red ball" and is constant throughout an episode but changes across episodes.  Each token can be encoded as a one-hot vector.  For simpler versions of the environment, all of the missions have the same structure (e.g. "go to the [color] [object]"), but larger ones can have a variable-size structure with multiple steps.

Reply:
Got it, nice